In [1]:
import pandas as pd

In [2]:
train_data_file = '/Users/mackbookair/Desktop/Git/CapStone/CapstoneProject/TestData/allstate-purchase-prediction-challenge-datawargling/train.csv'

In [3]:
# Load train data set into data frame

train_df = pd.read_csv(train_data_file)

In [4]:
train_df.head()

customer_ID  shopping_pt  record_type  day   time state  location  \
0     10000000            1            0    0  08:35    IN     10001   
1     10000000            2            0    0  08:38    IN     10001   
2     10000000            3            0    0  08:38    IN     10001   
3     10000000            4            0    0  08:39    IN     10001   
4     10000000            5            0    0  11:55    IN     10001   

   group_size  homeowner  car_age  ... C_previous  duration_previous  A  B  C  \
0           2          0        2  ...        1.0                2.0  1  0  2   
1           2          0        2  ...        1.0                2.0  1  0  2   
2           2          0        2  ...        1.0                2.0  1  0  2   
3           2          0        2  ...        1.0                2.0  1  0  2   
4           2          0        2  ...        1.0                2.0  1  0  2   

   D  E  F  G  cost  
0  2  1  2  2   633  
1  2  1  2  1   630  
2  2  1  2  1   630  
3  2  1  2  1   630  
4  2  1  2  1   630  

[5 rows x 25 columns]

In [5]:
# Inspect train data set
train_df.columns

Index(['customer_ID', 'shopping_pt', 'record_type', 'day', 'time', 'state',
       'location', 'group_size', 'homeowner', 'car_age', 'car_value',
       'risk_factor', 'age_oldest', 'age_youngest', 'married_couple',
       'C_previous', 'duration_previous', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
       'cost'],
      dtype='object')

In [6]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 665249 entries, 0 to 665248
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   customer_ID        665249 non-null  int64  
 1   shopping_pt        665249 non-null  int64  
 2   record_type        665249 non-null  int64  
 3   day                665249 non-null  int64  
 4   time               665249 non-null  object 
 5   state              665249 non-null  object 
 6   location           665249 non-null  int64  
 7   group_size         665249 non-null  int64  
 8   homeowner          665249 non-null  int64  
 9   car_age            665249 non-null  int64  
 10  car_value          663718 non-null  object 
 11  risk_factor        424831 non-null  float64
 12  age_oldest         665249 non-null  int64  
 13  age_youngest       665249 non-null  int64  
 14  married_couple     665249 non-null  int64  
 15  C_previous         646538 non-null  float64
 16  du

Observations:
1. 665249 records
2. 25 columns
3. time column is identified as object
3. Following columns have Null values -
    car_value,
    risk_factor,
    C_previous,
    duration_previous
4. column name 'customer_ID' is inconsistent


In [7]:
# changing the customer_ID column name to customer_id
train_df.rename(columns={'customer_ID':'customer_id'},inplace = True)

In [8]:
train_df.columns

Index(['customer_id', 'shopping_pt', 'record_type', 'day', 'time', 'state',
       'location', 'group_size', 'homeowner', 'car_age', 'car_value',
       'risk_factor', 'age_oldest', 'age_youngest', 'married_couple',
       'C_previous', 'duration_previous', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
       'cost'],
      dtype='object')

In [13]:
#dropping the na rows
# Is this better to drop na records or is it better to default ?
train_df_clean = train_df.dropna()
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   customer_id        417447 non-null  int64  
 1   shopping_pt        417447 non-null  int64  
 2   record_type        417447 non-null  int64  
 3   day                417447 non-null  int64  
 4   time               417447 non-null  object 
 5   state              417447 non-null  object 
 6   location           417447 non-null  int64  
 7   group_size         417447 non-null  int64  
 8   homeowner          417447 non-null  int64  
 9   car_age            417447 non-null  int64  
 10  car_value          417447 non-null  object 
 11  risk_factor        417447 non-null  float64
 12  age_oldest         417447 non-null  int64  
 13  age_youngest       417447 non-null  int64  
 14  married_couple     417447 non-null  int64  
 15  C_previous         417447 non-null  float64
 16  du

Record count decresed from 665,242 to 417,447

In [14]:
# Drop duplicate records
train_df_clean = train_df_clean.drop_duplicates()
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   customer_id        417447 non-null  int64  
 1   shopping_pt        417447 non-null  int64  
 2   record_type        417447 non-null  int64  
 3   day                417447 non-null  int64  
 4   time               417447 non-null  object 
 5   state              417447 non-null  object 
 6   location           417447 non-null  int64  
 7   group_size         417447 non-null  int64  
 8   homeowner          417447 non-null  int64  
 9   car_age            417447 non-null  int64  
 10  car_value          417447 non-null  object 
 11  risk_factor        417447 non-null  float64
 12  age_oldest         417447 non-null  int64  
 13  age_youngest       417447 non-null  int64  
 14  married_couple     417447 non-null  int64  
 15  C_previous         417447 non-null  float64
 16  du

No duplicates were observed.

Will inspect individual columns.

In [15]:
train_df_clean.risk_factor.unique()

array([3., 4., 2., 1.])

In [16]:
# Converting rist_factor data type to int
train_df_clean['risk_factor'] = train_df_clean['risk_factor'].astype('int')
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   customer_id        417447 non-null  int64  
 1   shopping_pt        417447 non-null  int64  
 2   record_type        417447 non-null  int64  
 3   day                417447 non-null  int64  
 4   time               417447 non-null  object 
 5   state              417447 non-null  object 
 6   location           417447 non-null  int64  
 7   group_size         417447 non-null  int64  
 8   homeowner          417447 non-null  int64  
 9   car_age            417447 non-null  int64  
 10  car_value          417447 non-null  object 
 11  risk_factor        417447 non-null  int64  
 12  age_oldest         417447 non-null  int64  
 13  age_youngest       417447 non-null  int64  
 14  married_couple     417447 non-null  int64  
 15  C_previous         417447 non-null  float64
 16  du

In [16]:
train_df_clean.C_previous.unique()

array([1., 3., 2., 4.])

In [18]:
# Converting C_previous column to int type

train_df_clean['C_previous'] = train_df_clean['C_previous'].astype('int')
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   customer_id        417447 non-null  int64  
 1   shopping_pt        417447 non-null  int64  
 2   record_type        417447 non-null  int64  
 3   day                417447 non-null  int64  
 4   time               417447 non-null  object 
 5   state              417447 non-null  object 
 6   location           417447 non-null  int64  
 7   group_size         417447 non-null  int64  
 8   homeowner          417447 non-null  int64  
 9   car_age            417447 non-null  int64  
 10  car_value          417447 non-null  object 
 11  risk_factor        417447 non-null  int64  
 12  age_oldest         417447 non-null  int64  
 13  age_youngest       417447 non-null  int64  
 14  married_couple     417447 non-null  int64  
 15  C_previous         417447 non-null  int64  
 16  du

In [18]:
train_df_clean.duration_previous.unique()

array([ 2., 13.,  3.,  0.,  1., 14.,  5., 15.,  4.,  6., 11.,  9., 10.,
        8.,  7., 12.])

In [19]:
# Converting duration_previous column to int type

train_df_clean['duration_previous']=train_df_clean['duration_previous'].astype('int')
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   customer_id        417447 non-null  int64 
 1   shopping_pt        417447 non-null  int64 
 2   record_type        417447 non-null  int64 
 3   day                417447 non-null  int64 
 4   time               417447 non-null  object
 5   state              417447 non-null  object
 6   location           417447 non-null  int64 
 7   group_size         417447 non-null  int64 
 8   homeowner          417447 non-null  int64 
 9   car_age            417447 non-null  int64 
 10  car_value          417447 non-null  object
 11  risk_factor        417447 non-null  int64 
 12  age_oldest         417447 non-null  int64 
 13  age_youngest       417447 non-null  int64 
 14  married_couple     417447 non-null  int64 
 15  C_previous         417447 non-null  int64 
 16  duration_previous  4

Inspecting other columns

In [30]:
train_df_clean.A.unique()

array([1, 0, 2])

In [21]:
len(train_df_clean.A.unique())

3

In [22]:
train_df_clean.B.unique()

array([0, 1])

In [23]:
train_df_clean.C.unique()

array([2, 3, 1, 4])

In [24]:
train_df_clean.D.unique()

array([2, 3, 1])

In [25]:
train_df_clean.E.unique()

array([1, 0])

In [26]:
train_df_clean.F.unique()

array([2, 0, 1, 3])

In [27]:
train_df_clean.G.unique()

array([2, 1, 3, 4])

In [28]:
train_df_clean.C_previous.unique()

array([1, 3, 2, 4])

In [29]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   customer_id        417447 non-null  int64 
 1   shopping_pt        417447 non-null  int64 
 2   record_type        417447 non-null  int64 
 3   day                417447 non-null  int64 
 4   time               417447 non-null  object
 5   state              417447 non-null  object
 6   location           417447 non-null  int64 
 7   group_size         417447 non-null  int64 
 8   homeowner          417447 non-null  int64 
 9   car_age            417447 non-null  int64 
 10  car_value          417447 non-null  object
 11  risk_factor        417447 non-null  int64 
 12  age_oldest         417447 non-null  int64 
 13  age_youngest       417447 non-null  int64 
 14  married_couple     417447 non-null  int64 
 15  C_previous         417447 non-null  int64 
 16  duration_previous  4

In [31]:
train_df_clean['car_value']

0         g
1         g
2         g
3         g
4         g
         ..
665232    g
665233    g
665240    f
665241    f
665242    g
Name: car_value, Length: 417447, dtype: object

In [32]:
train_df_clean.car_value.unique()

array(['g', 'e', 'd', 'f', 'c', 'h', 'i', 'b', 'a'], dtype=object)

In [33]:
train_df_clean['car_value'].value_counts()

e    135026
f    113445
d     69283
g     63481
h     20106
c     11911
i      2867
b       712
a       616
Name: car_value, dtype: int64

In [34]:

train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   customer_id        417447 non-null  int64   
 1   shopping_pt        417447 non-null  int64   
 2   record_type        417447 non-null  int64   
 3   day                417447 non-null  int64   
 4   time               417447 non-null  object  
 5   state              417447 non-null  object  
 6   location           417447 non-null  int64   
 7   group_size         417447 non-null  int64   
 8   homeowner          417447 non-null  int64   
 9   car_age            417447 non-null  int64   
 10  car_value          417447 non-null  category
 11  risk_factor        417447 non-null  int64   
 12  age_oldest         417447 non-null  int64   
 13  age_youngest       417447 non-null  int64   
 14  married_couple     417447 non-null  int64   
 15  C_previous         417447 non-null

In [35]:
train_df_clean.state.unique()

array(['IN', 'NY', 'WV', 'MO', 'PA', 'OH', 'FL', 'OR', 'WA', 'OK', 'ID',
       'CO', 'AL', 'MS', 'NM', 'CT', 'MD', 'RI', 'ME', 'TN', 'KS', 'KY',
       'NV', 'NE', 'UT', 'ND', 'DE', 'GA', 'IA', 'AR', 'WI', 'MT', 'NH',
       'DC', 'WY', 'SD'], dtype=object)

In [36]:
train_df_clean['state'].value_counts()

FL    103762
NY     89334
MD     27557
PA     26194
IN     23675
OH     15656
WA     13101
CO     10976
AL      9883
OR      7793
TN      7193
CT      7139
NV      6248
ID      5748
UT      5603
MO      5516
KY      5228
OK      4728
WI      4265
GA      4028
DE      3972
MS      3411
NH      3277
AR      3154
RI      3095
NM      2999
KS      2442
ME      2275
WV      2080
IA      1697
NE      1604
MT      1511
DC      1078
WY       563
ND       456
SD       206
Name: state, dtype: int64

In [37]:
train_df_clean['day'].value_counts()

0    90026
1    86350
2    83883
3    76308
4    74723
5     5442
6      715
Name: day, dtype: int64

In [38]:
train_df_clean['location'].value_counts()

10213    974
11517    864
10030    781
12091    730
10793    693
        ... 
14702      1
16026      1
15220      1
15862      1
15199      1
Name: location, Length: 5888, dtype: int64

In [39]:
train_df_clean['group_size'].value_counts()

1    320197
2     90918
3      5893
4       439
Name: group_size, dtype: int64

In [40]:
train_df_clean['homeowner'].value_counts()

1    248041
0    169406
Name: homeowner, dtype: int64

In [41]:
train_df_clean['car_age'].value_counts()

1     50441
2     34040
7     28664
6     28180
8     27678
      ...  
85        4
59        3
51        2
62        1
53        1
Name: car_age, Length: 66, dtype: int64

In [42]:
train_df_clean['car_value'].value_counts()

e    135026
f    113445
d     69283
g     63481
h     20106
c     11911
i      2867
b       712
a       616
Name: car_value, dtype: int64

In [43]:
train_df_clean['risk_factor'].value_counts()

3    114562
4    108632
1     98472
2     95781
Name: risk_factor, dtype: int64

In [44]:
train_df_clean['age_oldest'].value_counts()

75    40029
25    11944
24    11232
23    10687
26    10645
22     9208
27     9069
28     8461
66     7678
29     7555
55     7501
65     7500
63     7198
30     7133
58     7057
56     7027
59     7016
31     6986
60     6845
62     6830
32     6788
61     6732
70     6717
64     6714
52     6679
49     6675
54     6631
21     6601
53     6597
33     6581
57     6543
51     6478
48     6374
43     6326
50     6171
46     6160
42     6133
67     6088
69     6087
34     5995
44     5901
35     5864
45     5843
71     5786
41     5619
47     5585
68     5503
37     5467
36     5412
40     5396
39     5171
72     5151
38     5083
74     4945
73     4888
20     4582
19     3426
18     1154
Name: age_oldest, dtype: int64

In [45]:
train_df_clean['age_youngest'].value_counts()

75    30795
25    13303
24    12458
23    12124
26    11768
22    10750
27    10053
28     9567
29     8373
21     8364
30     7856
31     7801
32     7412
65     7292
66     7122
55     7043
33     6856
58     6627
62     6558
63     6545
60     6536
56     6453
61     6384
34     6349
35     6327
20     6322
64     6302
59     6243
52     6236
57     6223
54     6155
53     6116
51     6110
49     6059
70     5997
48     5924
42     5806
69     5756
50     5745
46     5732
43     5726
44     5552
37     5527
45     5520
41     5406
36     5396
71     5380
40     5366
38     5320
19     5296
68     5224
39     5186
47     5104
67     5050
72     4463
73     4451
74     4422
18     3149
16     2676
17     1791
Name: age_youngest, dtype: int64

In [46]:
train_df_clean['married_couple'].value_counts()

0    322944
1     94503
Name: married_couple, dtype: int64

In [47]:
train_df_clean['duration_previous'].value_counts()

15    64736
2     44079
1     42179
3     41620
4     35712
5     32225
6     30164
7     26330
8     21591
9     19175
10    13765
0     12800
11     9480
12     8204
13     8149
14     7238
Name: duration_previous, dtype: int64

In [48]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   customer_id        417447 non-null  int64   
 1   shopping_pt        417447 non-null  int64   
 2   record_type        417447 non-null  int64   
 3   day                417447 non-null  int64   
 4   time               417447 non-null  object  
 5   state              417447 non-null  object  
 6   location           417447 non-null  int64   
 7   group_size         417447 non-null  int64   
 8   homeowner          417447 non-null  int64   
 9   car_age            417447 non-null  int64   
 10  car_value          417447 non-null  category
 11  risk_factor        417447 non-null  int64   
 12  age_oldest         417447 non-null  int64   
 13  age_youngest       417447 non-null  int64   
 14  married_couple     417447 non-null  int64   
 15  C_previous         417447 non-null

In [51]:
# Converting car_value, married_couple and state columns to category

train_df_clean['married_couple'] = train_df_clean['married_couple'].astype('category')
train_df_clean['state'] = train_df_clean['state'].astype('category')
train_df_clean['car_value'] = train_df_clean['car_value'].astype('category')

print(train_df_clean.dtypes)

customer_id             int64
shopping_pt             int64
record_type             int64
day                     int64
time                   object
state                category
location                int64
group_size              int64
homeowner               int64
car_age                 int64
car_value            category
risk_factor             int64
age_oldest              int64
age_youngest            int64
married_couple       category
C_previous              int64
duration_previous       int64
A                       int64
B                       int64
C                       int64
D                       int64
E                       int64
F                       int64
G                       int64
cost                    int64
dtype: object


In [52]:
train_df_clean['married_couple']

0         1
1         1
2         1
3         1
4         1
         ..
665232    0
665233    0
665240    0
665241    0
665242    0
Name: married_couple, Length: 417447, dtype: category
Categories (2, int64): [0, 1]

In [53]:
train_df_clean['state']

0         IN
1         IN
2         IN
3         IN
4         IN
          ..
665232    FL
665233    FL
665240    FL
665241    FL
665242    FL
Name: state, Length: 417447, dtype: category
Categories (36, object): [AL, AR, CO, CT, ..., WA, WI, WV, WY]

In [54]:
train_df_clean['car_value']

0         g
1         g
2         g
3         g
4         g
         ..
665232    g
665233    g
665240    f
665241    f
665242    g
Name: car_value, Length: 417447, dtype: category
Categories (9, object): [a, b, c, d, ..., f, g, h, i]

In [55]:
# checking for successful purchase records
len(train_df_clean[train_df_clean.record_type==1])

62255

In [56]:
# cross checking all successful purchase records are unique
len(train_df_clean[train_df_clean.record_type==1].customer_id.unique())

62255

In [57]:
len(train_df_clean.customer_id.unique())

62978

In [57]:
#l = train_df_clean[train_df_clean.record_type==1]
#l

only 62255 records have purchased coverage options out of 62978 unique customer_ids

In [58]:
# fetching only records that have a purchased coverage
# Is this a good approach for this data set, to drop those records which do not have a record_type as 1 ?

train_df_p = train_df_clean[train_df_clean['customer_id'].isin(train_df_clean[train_df_clean.record_type==1].customer_id)]
train_df_p

customer_id  shopping_pt  record_type  day   time state  location  \
0          10000000            1            0    0  08:35    IN     10001   
1          10000000            2            0    0  08:38    IN     10001   
2          10000000            3            0    0  08:38    IN     10001   
3          10000000            4            0    0  08:39    IN     10001   
4          10000000            5            0    0  11:55    IN     10001   
...             ...          ...          ...  ...    ...   ...       ...   
665232     10152720            7            0    3  15:24    FL     14100   
665233     10152720            8            1    4  08:34    FL     14100   
665240     10152723            1            0    2  15:13    FL     10711   
665241     10152723            2            0    2  15:14    FL     10711   
665242     10152723            3            1    1  10:30    FL     10711   

        group_size  homeowner  car_age  ... C_previous  duration_previous  A  \
0                2          0        2  ...          1                  2  1   
1                2          0        2  ...          1                  2  1   
2                2          0        2  ...          1                  2  1   
3                2          0        2  ...          1                  2  1   
4                2          0        2  ...          1                  2  1   
...            ...        ...      ...  ...        ...                ... ..   
665232           1          0       13  ...          1                  8  1   
665233           1          0       13  ...          1                  8  1   
665240           1          1        0  ...          3                  2  0   
665241           1          1        0  ...          3                  2  1   
665242           1          1        0  ...          3                  7  1   

        B  C  D  E  F  G  cost  
0       0  2  2  1  2  2   633  
1       0  2  2  1  2  1   630  
2       0  2  2  1  2  1   630  
3       0  2  2  1  2  1   630  
4       0  2  2  1  2  1   630  
...    .. .. .. .. .. ..   ...  
665232  0  1  2  0  2  3   681  
665233  0  1  2  0  2  3   681  
665240  0  2  1  0  0  4   656  
665241  0  3  3  0  2  3   687  
665242  0  3  3  1  2  3   651  

[414812 rows x 25 columns]

In [59]:
len(train_df_p.customer_id.unique())

62255

In [60]:
train_df_p.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414812 entries, 0 to 665242
Data columns (total 25 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   customer_id        414812 non-null  int64   
 1   shopping_pt        414812 non-null  int64   
 2   record_type        414812 non-null  int64   
 3   day                414812 non-null  int64   
 4   time               414812 non-null  object  
 5   state              414812 non-null  category
 6   location           414812 non-null  int64   
 7   group_size         414812 non-null  int64   
 8   homeowner          414812 non-null  int64   
 9   car_age            414812 non-null  int64   
 10  car_value          414812 non-null  category
 11  risk_factor        414812 non-null  int64   
 12  age_oldest         414812 non-null  int64   
 13  age_youngest       414812 non-null  int64   
 14  married_couple     414812 non-null  category
 15  C_previous         414812 non-null

In [62]:
# Droping time column

train_df_final = train_df_p.drop('time',axis=1)
train_df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414812 entries, 0 to 665242
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   customer_id        414812 non-null  int64   
 1   shopping_pt        414812 non-null  int64   
 2   record_type        414812 non-null  int64   
 3   day                414812 non-null  int64   
 4   state              414812 non-null  category
 5   location           414812 non-null  int64   
 6   group_size         414812 non-null  int64   
 7   homeowner          414812 non-null  int64   
 8   car_age            414812 non-null  int64   
 9   car_value          414812 non-null  category
 10  risk_factor        414812 non-null  int64   
 11  age_oldest         414812 non-null  int64   
 12  age_youngest       414812 non-null  int64   
 13  married_couple     414812 non-null  category
 14  C_previous         414812 non-null  int64   
 15  duration_previous  414812 non-null

In [63]:
#categorical data
categorical_cols = ['state', 'car_value','married_couple'] 
from sklearn.preprocessing import LabelEncoder
# instantiate labelencoder object
le = LabelEncoder()
# apply le on categorical feature columns
train_df_final[categorical_cols] = train_df_final[categorical_cols].apply(lambda col: le.fit_transform(col))
train_df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414812 entries, 0 to 665242
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype
---  ------             --------------   -----
 0   customer_id        414812 non-null  int64
 1   shopping_pt        414812 non-null  int64
 2   record_type        414812 non-null  int64
 3   day                414812 non-null  int64
 4   state              414812 non-null  int64
 5   location           414812 non-null  int64
 6   group_size         414812 non-null  int64
 7   homeowner          414812 non-null  int64
 8   car_age            414812 non-null  int64
 9   car_value          414812 non-null  int64
 10  risk_factor        414812 non-null  int64
 11  age_oldest         414812 non-null  int64
 12  age_youngest       414812 non-null  int64
 13  married_couple     414812 non-null  int64
 14  C_previous         414812 non-null  int64
 15  duration_previous  414812 non-null  int64
 16  A                  414812 non-null  in

In [64]:
train_df_final['record_type'].value_counts()

0    352557
1     62255
Name: record_type, dtype: int64

In [65]:
#df.drop(df.query('col1 < 6').sample(frac=.4).index)
#df.loc[df['column_name'] == some_value]
train_df_final_RandomSample = train_df_final.drop(train_df_final.loc[train_df_final['record_type']==0].sample(frac=.8).index) 

In [66]:
train_df_final_RandomSample.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 132766 entries, 3 to 665242
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype
---  ------             --------------   -----
 0   customer_id        132766 non-null  int64
 1   shopping_pt        132766 non-null  int64
 2   record_type        132766 non-null  int64
 3   day                132766 non-null  int64
 4   state              132766 non-null  int64
 5   location           132766 non-null  int64
 6   group_size         132766 non-null  int64
 7   homeowner          132766 non-null  int64
 8   car_age            132766 non-null  int64
 9   car_value          132766 non-null  int64
 10  risk_factor        132766 non-null  int64
 11  age_oldest         132766 non-null  int64
 12  age_youngest       132766 non-null  int64
 13  married_couple     132766 non-null  int64
 14  C_previous         132766 non-null  int64
 15  duration_previous  132766 non-null  int64
 16  A                  132766 non-null  in

In [67]:
#chekcing for the number of record_type records with 0 and 1
train_df_final_RandomSample['record_type'].value_counts()

0    70511
1    62255
Name: record_type, dtype: int64

Now the samples for record_type 0 and 1 are about same and it gives the model enough data to learn on both type of records.

In [68]:
y_recordtype = train_df_final_RandomSample['record_type']
type(y_recordtype)

pandas.core.series.Series

In [69]:
y_recordtype

3         0
8         1
14        1
26        1
27        0
         ..
665222    0
665225    1
665228    0
665233    1
665242    1
Name: record_type, Length: 132766, dtype: int64

In [72]:
X_df = train_df_final_RandomSample.drop('record_type',axis=1)
X_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 132766 entries, 3 to 665242
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype
---  ------             --------------   -----
 0   customer_id        132766 non-null  int64
 1   shopping_pt        132766 non-null  int64
 2   day                132766 non-null  int64
 3   state              132766 non-null  int64
 4   location           132766 non-null  int64
 5   group_size         132766 non-null  int64
 6   homeowner          132766 non-null  int64
 7   car_age            132766 non-null  int64
 8   car_value          132766 non-null  int64
 9   risk_factor        132766 non-null  int64
 10  age_oldest         132766 non-null  int64
 11  age_youngest       132766 non-null  int64
 12  married_couple     132766 non-null  int64
 13  C_previous         132766 non-null  int64
 14  duration_previous  132766 non-null  int64
 15  A                  132766 non-null  int64
 16  B                  132766 non-null  in

In [73]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Split the data into a training and test set.
Xlr, Xtestlr, ylr, ytestlr = train_test_split(X_df, 
                                              y_recordtype, test_size=0.4,random_state=42)

In [74]:
clf = LogisticRegression()
# Fit the model on the trainng data.
clf.fit(Xlr, ylr)
# Print the accuracy from the testing data.
print(accuracy_score(clf.predict(Xtestlr), ytestlr))

0.5315683431562694


In [75]:
print(ytestlr.value_counts())

0    28230
1    24877
Name: record_type, dtype: int64


In [76]:
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [77]:
print(confusion_matrix(ytestlr, clf.predict(Xtestlr)))

[[28230     0]
 [24877     0]]


In [78]:
print(classification_report(ytestlr, clf.predict(Xtestlr)))

              precision    recall  f1-score   support

           0       0.53      1.00      0.69     28230
           1       0.00      0.00      0.00     24877

    accuracy                           0.53     53107
   macro avg       0.27      0.50      0.35     53107
weighted avg       0.28      0.53      0.37     53107



/Applications/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [79]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=6)

In [80]:
knn.fit(Xlr,ylr)
y_pred_knn = knn.predict(Xtestlr)

In [81]:
print(confusion_matrix(ytestlr, y_pred_knn))

[[20506  7724]
 [20112  4765]]


In [82]:
print(classification_report(ytestlr, y_pred_knn))

              precision    recall  f1-score   support

           0       0.50      0.73      0.60     28230
           1       0.38      0.19      0.26     24877

    accuracy                           0.48     53107
   macro avg       0.44      0.46      0.43     53107
weighted avg       0.45      0.48      0.44     53107



In [83]:
y_pred_knn

array([0, 1, 0, ..., 0, 0, 0])